base/checked.jl

```diff
@@ -0,0 +1,333 @@
+export checked_neg, checked_abs, checked_add, checked_sub, checked_mul,
+       checked_div, checked_rem, checked_fld, checked_mod, checked_cld
+
+import Core.Intrinsics: box, unbox,
+       checked_sadd_int, checked_ssub_int, checked_smul_int, checked_sdiv_int,
+       checked_srem_int,
+       checked_uadd_int, checked_usub_int, checked_umul_int, checked_udiv_int,
+       checked_urem_int
+
+typealias SignedInt Union{Int8,Int16,Int32,Int64,Int128}
+typealias UnsignedInt Union{UInt8,UInt16,UInt32,UInt64,UInt128}
+
+# LLVM has several code generation bugs for checked integer arithmetic (see e.g.
+# #4905). We thus distinguish between operations that can be implemented via
+# intrinsics, and operations for which we have to provide work-arounds.
+
+# Note: As far as this code has been tested, most checked_* functions are
+# working fine in LLVM. (Note that division is still handled via `base/int.jl`,
+# which always checks for overflow, and which provides its own sets of
+# work-arounds for LLVM codegen bugs.) However, the comments in `base/int.jl`
+# and in issue #4905 are more pessimistic. For the time being, we thus retain
+# the ability to handle codegen bugs in LLVM, until the code here has been
+# tested on more systems and architectures. It also seems that things depend on
+# which compiler that was used to build LLVM (i.e. either gcc or clang).
+
+# These unions are used for most checked functions:
+#     BrokenSignedInt
+#     BrokenUnsignedInt
+# These unions are used for checked_{mul,div,rem}:
+#     BrokenSignedIntMul
+#     BrokenUnsignedIntMul
+
+# This code runs early during bootstrap, and we can't use Julia's version
+# strings yet
+const llvm_version = Int(ccall(:jl_get_LLVM_VERSION, UInt32, ()))
+if llvm_version >= 30500
+    if WORD_SIZE == 32
+        typealias BrokenSignedInt Union{}
+        typealias BrokenUnsignedInt Union{}
+        typealias BrokenSignedIntMul Union{Int64,Int128}
+        typealias BrokenUnsignedIntMul Union{UInt64,UInt128}
+    else
+        typealias BrokenSignedInt Union{}
+        typealias BrokenUnsignedInt Union{}
+        typealias BrokenSignedIntMul Int128
+        typealias BrokenUnsignedIntMul UInt128
+    end
+else
+    if WORD_SIZE == 32
+        typealias BrokenSignedInt Union{}
+        typealias BrokenUnsignedInt Union{}
+        typealias BrokenSignedIntMul Union{Int8,Int64,Int128}
+        typealias BrokenUnsignedIntMul Union{UInt8,UInt64,UInt128}
+    else
+        typealias BrokenSignedInt Union{}
+        typealias BrokenUnsignedInt Union{}
+        typealias BrokenSignedIntMul Union{Int8,Int128}
+        typealias BrokenUnsignedIntMul Union{UInt8,UInt128}
+    end
+end
+# Use these definitions to test the non-LLVM implementations
+# typealias BrokenSignedInt SignedInt
+# typealias BrokenUnsignedInt UnsignedInt
+# typealias BrokenSignedIntMul SignedInt
+# typealias BrokenUnsignedIntMul UnsignedInt
+
+# A tight `widen`
+nextwide(::Type{Int8}) = Int16
```

Comment:
`widen` jumps right away to `Int64`, which, if you are going to be allocating an array, would be very wasteful.